src/hotspot/share/gc/g1/heapRegion.cpp

```diff
@@ -483,125 +483,169 @@ static bool is_oop_safe(oop obj) {
 // Closure that glues together validity check for oop references (first),
 // then optionally verifies the remembered set for that reference.
 class G1VerifyLiveAndRemSetClosure : public BasicOopIterateClosure {
-  using CardValue = CardTable::CardValue;
-
-  G1CollectedHeap* _g1h;
   VerifyOption _vo;
-
   oop _containing_obj;
-
   size_t _num_failures;
 
-  G1CardTable *_ct;
-
-  bool has_failures() const { return _num_failures != 0; }
+  // Increases the failure counter and return whether this has been the first failure.
+  bool record_failure() {
+    _num_failures++;
+    return _num_failures == 1;
+  }
 
-  void print_object(outputStream* out, oop obj) {
+  static void print_object(outputStream* out, oop obj) {
 #ifdef PRODUCT
     obj->print_name_on(out);
 #else // PRODUCT
     obj->print_on(out);
 #endif // PRODUCT
   }
 
-  template<class T>
-  bool verify_liveness(T* p, oop obj) {
-    bool is_in_heap = _g1h->is_in(obj);
+  template <class T>
+  class Checker {
+  public:
+    G1CollectedHeap* _g1h;
+    G1VerifyLiveAndRemSetClosure* _cl;
+    oop _containing_obj;
+    T* _p;
+    oop _obj;
+
+    Checker(G1VerifyLiveAndRemSetClosure* cl, oop containing_obj, T* p, oop obj) :
+      _g1h(G1CollectedHeap::heap()),
+      _cl(cl),
+      _containing_obj(containing_obj),
+      _p(p),
+      _obj(obj) { }
+
+    void print_containing_obj(outputStream* out, HeapRegion* from) {
+      log_error(gc, verify)("Field " PTR_FORMAT " of obj " PTR_FORMAT " in region " HR_FORMAT,
+                            p2i(_p), p2i(_containing_obj), HR_FORMAT_PARAMS(from));
+      print_object(out, _containing_obj);
+    }
+
+    void print_referenced_obj(outputStream* out, HeapRegion* to, const char* explanation) {
+      log_error(gc, verify)("points to %sobj " PTR_FORMAT " in region " HR_FORMAT " remset %s",
+                            explanation, p2i(_obj), HR_FORMAT_PARAMS(to), to->rem_set()->get_state_str());
+      print_object(out, _obj);
+    }
+  };
+
+  template <class T>
+  class LiveChecker : public Checker<T> {
+  public:
+    VerifyOption _vo;
+    bool _is_in_heap;
+
+    LiveChecker(G1VerifyLiveAndRemSetClosure* cl, oop containing_obj, T* p, oop obj, VerifyOption vo) : Checker<T>(cl, containing_obj, p, obj) {
+      _vo = vo;
+      _is_in_heap = this->_g1h->is_in(obj);
+    }
 
-    if (!is_in_heap || _g1h->is_obj_dead_cond(obj, _vo)) {
+    bool failed() const {
+      return !_is_in_heap || this->_g1h->is_obj_dead_cond(this->_obj, _vo);
```

Comment:
Not sure why `this->` is there. I think breaking the boundary btw child & super class is fine in this particular patch.

Reply:
`this->` is required in C++ when accessing the base-class members of inherited template classes. Without this nudge, the compiler does not look up `_g1h` and `_obj` in the template base class and gives a compilation error.